### DISCUSSION TITLE
`render()` performance improvement

### DISCUSSION THREAD

**Discussion by christoph-heinrich** [tomasklaen/uosc#255] [Ideas]
In Lua when concatenating a lot of strings with `a .. b` like we are doing, it is generally a lot faster to have a table, append each string to that table and at the end get the final string with `table.concat(strings)`.

For me rendering each element when everything is visible takes ~4ms, so there is no huge need for improvement.
Just wanted to put it out there in case someone is motivated enough to actually touch all of rendering.

Also in case anyone actually does want to tackle this, something like `a .. '\meh' .. b .. c` gets optimized and creates a single string, instead of concatenating each `..` operation individually, so I don't think it would be all that much faster (maybe even slower) to added each of those strings to the table individually. But that only happens when it's in a single expression.

In [Programming in Lua](https://www.lua.org/pil/11.6.html) they have a little algorithm to already concatenate things along the way, but I have no idea if that's actually faster or not.

**Comment by tomasklaen** [comment]
o.O I'd be surprised if string concatenation change would result in any noticeable difference. That has to be less than a single microsecond. 99% bottleneck here is definitely libass which is out of our control.

**Reply by christoph-heinrich** [reply]
That time was measured in Lua, so no libass here. Simply `mp.get_time()` after the loop and then subtract `render_last_time` from that.

**Reply by tomasklaen** [reply]
😨 that is actually surprising to me. Either way, majority of string concatenations are happening in [assdraw.lua](https://github.com/mpv-player/mpv/blob/master/player/lua/assdraw.lua). So unless you want to rewrite everything from the ground up, this is a pipe dream :)

**Reply by christoph-heinrich** [reply]
Out of curiosity I've now also checked how long libass takes in the same scenario. ~1.7ms for me.

> fearful that is actually surprising to me. Either way, majority of string concatenations are happening in [assdraw.lua](https://github.com/mpv-player/mpv/blob/master/player/lua/assdraw.lua). So unless you want to rewrite everything from the ground up, this is a pipe dream :)

Add a new API to assdraw.lua that accepts a string buffer :laughing:.

Although, if it kept the buffer internally, then the API wouldn't even have to change. Problem is that `ass.text` can always be accessed, which conflicts with the whole internal buffer idea.

**Reply by tomasklaen** [reply]
Wait, my render loop with toggle-ui on (top bar, volume, timeline, controls bar visible) takes from 0.3 to 0.6ms.

**Reply by christoph-heinrich** [reply]
I simply made the window small enough that everything gets shown (even if not fully opaque) through proximity. Should be the same amount of work for our code.
``` lua
print((mp.get_time() - state.render_last_time) * 1000)
```
![Screenshot_20220924_005734](https://user-images.githubusercontent.com/8932183/192067141-8c33ffc5-b524-4b3d-83a5-788858d454b4.png)

Btw my CPU is roughly 10 years old, so it's certainly possible that it runs that much faster on newer hardware :wink:

**Reply by tomasklaen** [reply]
I put the exact same print right before `osd:update()` line. And I'm on i5-6500, pretty old too.

**Reply by tomasklaen** [reply]
I've managed to achieve a slight increase in times by playing a stream:

![watch_v=Npu7jkJk5nM 00_00_41 341 0001](https://user-images.githubusercontent.com/47283320/192107519-fbb77a4f-2c88-402a-9be9-c74f9be24b00.png)

But still insignificant. I have no idea why your render loop is so slow. It happened to me once when I was putting `print()` statements in between time bookmarks, but I don't think that's happening here.

**Reply by christoph-heinrich** [reply]
I suspect you're using luajit while I'm using just lua.

**Reply by tomasklaen** [reply]
I'm on shichiro's builds here. Dunno what lua is inside.

**Reply by christoph-heinrich** [reply]
I'm getting mostly ~0.7ms times with luajit, so that's definitely it.

> I'm on shichiro's builds here. Dunno what lua is inside.

Looks like a windows build, and afaik windows builds pretty much all use luajit.

**Comment by christoph-heinrich** [comment]
I've implemented this [here](https://github.com/christoph-heinrich/mpv/tree/ass_table) by using a table for `ass.text` instead of a string and setting `__concat()` in the metatable.

Performance increase is pretty small (maybe 20% for both lua and luajit, I didn't take a long term average so it's hard to say), but the memory increase is much slower.

The problem with that implementation is that the volume slider writes some text on the screen instead of drawing those commands. No idea what's wrong there and it's the only thing I could find that doesn't work correctly.
